### ISSUE TITLE
Document best new practice w/ lodash 3.x

### ISSUE THREAD

**Issue by SlexAxton** [artsyca/ember-cli-lodash#3]
This isn't really your job, but it's the only place I could think of to look for answer.

I can't really figure out how to import the ES6 lodash nicely. I've tried some combination of `vendor` and `app.import` on the lodash.js file and haven't come across any working solution.

If anyone knows how to best do it, it could be nice to place under the deprecation notice.

**Comment by SlexAxton** [comment]
After some additional research, I think the es6 stuff is still blocked a little -- https://github.com/ember-cli/ember-cli/issues/2949

**Comment by artsyca** [comment]
Yes, I was imagining that you'd have to just `import lodash from 'bower_components/lodash/...'` and the shim would no longer be needed. Being able to just `import lodash from 'lodash'` would be nice but I don't think that syntax would be worth maintaining an entire addon.

**Comment by artsyca** [comment]
Also, if the es6 module didn't work, you could use the AMD module and include an entry in `Brocfile.js`

**Comment by jdalton** [comment]
I made the [lodash-es](https://www.npmjs.com/package/lodash-es) with Ember in mind so if there's changes or tweaks let me know and I'll tackle it.

**Comment by SlexAxton** [comment]
It's an ember problem, I think, for now at least.

**Comment by seawatts** [comment]
@jdalton after install lodash-es with `npm install --save-dev lodash-es` do you know how to load lodash into a test in ember?

I've tried 
* `import _ from 'node_modules/lodash-es/lodash';` 
* `import _ from 'lodash';`
* `import _ from `my-app/node_modules/lodash-es/lodash';`

**Comment by jdalton** [comment]
@seawatts I'm not familiar with the recommended Ember approach. I currently have the "jsnext:main" field in the package.json which points to the [lodash](https://github.com/lodash/lodash/blob/3.6.0-es/lodash.js) module.

**Comment by artsyca** [comment]
@jdalton I believe `import _ from 'npm:lodash';` does the trick.

This doesn't allow for importing individual functions as `import { reduce } from 'npm:lodash'` however.

**Comment by seawatts** [comment]
@artsyca Hmm, that doesn't seem to work for me.

**Comment by jdalton** [comment]
Let me know if I need to tweak anything. The lodash-es build exists for Ember so y'alls needs are top priority for it.

**Comment by seawatts** [comment]
Thanks @jdalton I'll keep you updated.

@artsyca do you think you have to explicitly add it like this http://www.ember-cli.com/#test-assets

**Comment by SlexAxton** [comment]
So the current problem is that ember-cli doesn't support transpiling vendor stuff - so the es6 branch can't be used directly (at least as a bower/npm dep). So it's more likely that the AMD branch is actually the one that would work best, but as far as I know it might require us to enumerate the exportable values, which is silly.

I've been using a hack to just shim the global and then export the enumerable properties, which worked *alright* but would be cool to get the real deal working.

/cc @rwjblue @stefanpenner - any great ideas on how to get the es6 lodash working in ember-cli? Happy to do the leg-work.

**Comment by SlexAxton** [comment]
I would expect that the following worked:

```
// bower.json
'lodash': '3.6.0-amd',
```

```js
// Brocfile.js

// Bower as destructurable imports
app.import('bower_components/lodash/main.js', {
  type: 'vendor',
  exports: { 'lodash': ['default'] }
});
```

But it does not end up registered.

**Comment by SlexAxton** [comment]
OK. I tried modifying the `main.js` file a bit to see what would make things happier and found a solution that's silly, but works (so hopefully can point towards a better solution).

My Brocile is the same:

```js
app.import('bower_components/lodash/main.js', {
  type: 'vendor',
  exports: { 'lodash': ['default'] }
});
```

I killed the AMD check:

```js
if (typeof define == 'function' && typeof define.amd == 'object' && define.amd) {
```

This never evaluates to true when we're importing this file. Then I named the export, and gave it a non-shorthand signature:


```js
define('lodash', [], function() {
  return _;
});
```

Then in my regular ember modules, I could do this:

```js
import {forIn} from 'lodash';

forIn(foo, bar);
```

I think I might try to run the es6 version through a prefilter though, so they all just register without this hackery.

**Comment by SlexAxton** [comment]
Hey @jdalton -- working on getting the es version of lodash compiled. I ran into a quick parsing error.

This one: https://github.com/esperantojs/esperanto/blob/b5c875fdf753ea514ffe9381e230241126d987c7/src/utils/ast/annotate.js#L106

It's specifically mad about this line: https://github.com/lodash/lodash/blob/es/internal/root.js#L28

More specifically, the `|| this` at the end is ok, but the `freeWindow !== (this && this.window)` is making things angry. I can compile lodash if I change that line to be:

`var root = freeGlobal || freeWindow || freeSelf || this;`

but clearly I'm losing some compatibility if I do that.

Suggestions?

**Comment by jdalton** [comment]
Sounds like an esperanto bug? lodash-es is tested to transpile successfully with babel.js & es6-module-transpiler (I know it's an oldy).

**Comment by SlexAxton** [comment]
Perhaps? It seems like maybe they're following a strict mode rule that the others weren't following (and modules force strict mode). I would use babel, the ember-cli guys suggested their broccoli plugin for compat -- https://github.com/ember-cli/broccoli-es6modules

Based on the only threads I can find on the issue, it fails because `this` is `undefined` in these cases. So most libs literally just replace `this` with `undefined` - but since we have a property access `this.window` that fails ( as `undefined.window` ).

Either way, I think I've sunk enough time into this for now :D -- for anyone who ends up here, this is what I did:

* forked lodash - checked out `3.6.0-es`
* removed the `this` incompatibility by hand, as shown in the comment above.
* built it with https://github.com/ember-cli/broccoli-es6modules as a `namedAmd` output.
* committed the output to a branch -- https://github.com/SlexAxton/lodash/tree/slexaxton-3.6.0-es-ember-dist
* bower installed my branch --  `"lodash": "SlexAxton/lodash#slexaxton-3.6.0-es-ember-dist"`
* imported the built file as an AMD dep in my Brocfile.js

```js
// For lodash
app.import('bower_components/lodash/prebuilt-as-amd/lodash.js', {
  vendor: true,
  exports: {
    lodash: ['default']
  }
});
```

Now I can do this:

```js
import forIn from 'lodash';

forIn({a:1, b: 2}, function(val, key) {
  console.log(key, val);
});
```

--

The only downside to this (and I think even if this worked without a fork), is that we're still including *all* of lodash - even though we know our exact dependency tree.

Next time I work on this, I think I'll see if I can work out causing automatic builds from your actual used dep trees, though the only way I can think to do this is via importing from files:

```js
import forIn from 'lodash/forIn';
```

And then I usually run into some context and available "global lodash properties" - but I think it's possible to eventually get there.

**Comment by jdalton** [comment]
> Based on the only threads I can find on the issue, it fails because this is undefined in these cases. So most libs literally just replace this with undefined - but since we have a property access `this.window` that fails ( as `undefined.window` ).

Thanks for digging. babel.js adds the "use strict" directive when transpiling so if there was a strict mode error it would catch that (it has in the past). The `this` reference isn't an issue because it's not just `this.window` it's `(this && this.window)`. Even if it was transpiled to `(undefined && undefined.window)` it wouldn't error. Pinging @eventualbuddha and @Rich-Harris to see if `this` is a known issue.

**Comment by Rich-Harris** [comment]
Yeah, we thought it would save people pain if we treated top-level `this` in an ES6 module as a parse-time error (rather than just replacing it with `undefined` and waiting for things to blow up at runtime). Apparently that's not the case. Sorry @SlexAxton! [Discussion here](https://github.com/esperantojs/esperanto/pull/43), [PR here](https://github.com/esperantojs/esperanto/pull/56).

We could conceivably add an `allowThis` option to esperanto. Would that help solve this problem?

**Comment by Rich-Harris** [comment]
Having said that... this Greasemonkey stuff is one hell of an edge case. Is there a workaround that doesn't involved reading properties of `this`?

**Comment by jdalton** [comment]
> Yeah, we thought it would save people pain if we treated top-level this in an ES6 module as a parse-time error (rather than just replacing it with undefined and waiting for things to blow up at runtime). 

Ya a parse error seems aggressive. It's valid JS even in ES6 it's just has an `undefined` value.

**Comment by Rich-Harris** [comment]
Have opened a corresponding issue on esperanto - https://github.com/esperantojs/esperanto/issues/151

**Comment by artsyca** [comment]
Guys, I have to admit I'm a bit lost in this thread as there are multiple issues being discussed?

To address the issue of importing lodash: In our production project, we're including lodash as an npm dependency (not bower!) and import as `import _ from 'npm:lodash'` wherever we please. This is the reason I've discontinued the `ember-cli-lodash` project because now it just works out of the box (minus granular imports)

Apologies if I can't contribute much further to the conversation beyond that.

**Comment by seawatts** [comment]
@artsyca I'm also starting to get lost. =/ 

I have tried to do that as well, without luck. I am also trying to load it in an acceptance test. I wonder if thats different?

I'm using:

lodash: 3.6.0 - **installed with  ```npm install --save lodash``` not the es version**
ember-cli: 0.2.3
ember: 1.11.1
node: 0.12.2

**Comment by artsyca** [comment]
@seawatts : I'm so sorry, I left out a very important detail. I just consulted with my team, and the reason this works is because of [ember-browserify](https://github.com/ef4/ember-browserify).

**Comment by seawatts** [comment]
No big deal. This is interesting. I'll look into it. Thanks!

**Comment by Restuta** [comment]
So issue still remains, what is the default idiomatic way of using lodash ES6 module in Ember (without ember-browserify)

**Comment by SlexAxton** [comment]
There was some fix from @jdalton like 14 days ago for the related issue. I
think maybe it hasn't seen a release yet. JDD?

**Comment by jdalton** [comment]
esparanto bumped to 0.7.3 15 days go which fixed their issue. I'm due for a patch release but not sure if it fixes anything related to this.

**Comment by jdalton** [comment]
\cc @truenorth

**Comment by SlexAxton** [comment]
Thanks for the info, jdd. Hopefully a bump would fix it all. I'll check against masters here in a day or so.

**Comment by truenorth** [comment]
Still need a new ember-cli release to bring in [the new broccoli-es6-modules](https://github.com/ember-cli/broccoli-es6modules/releases/tag/v1.0.1) version. Slowly but surely, we're getting there!
